File: src/script/geocoding.rb
require 'dotenv'
require 'net/http'

Dotenv.load 'src/.env.local'
API_KEY = ENV['GOOGLE_GEOCODING_API_KEY']

def fetch_address(station)
  print "get address of station:#{station['name']} > "

  uri = URI.parse('https://maps.googleapis.com/')
  https = Net::HTTP.new(uri.host, uri.port)
  https.use_ssl = true
  https.verify_mode = OpenSSL::SSL::VERIFY_NONE
  https.verify_depth = 5
  res = https.start do |w|
    w.get("/maps/api/geocode/json?latlng=#{station['lat']},#{station['lng']}&key=#{API_KEY}&language=ja")
  end
  assert_equal res.code, '200', 'response from /maps/api/geocode/json'
  data = JSON.parse(res.body)
  assert_equal data['status'], 'OK', "response:\n#{JSON.pretty_generate(data)}"

  data = data['results'][0]

  puts "address: #{data['formatted_address']} #{data}"
  # 郵便番号
  list = data['address_components'].select { |c| c['types'].include?('postal_code') }
  assert_equal list.length, 1, 'fail to extract postal-code'
  station['postal_code'] = list[0]['long_name']
  # 住所
  exception = %w[postal_code country bus_station train_station transit_station]
  address = ''
  previous = nil
  pattern = /^[0-9０-９]+$/
  components = data['address_components']
  components.reject! do |c|
    c['types'].any? { |e| exception.include? e }
  end
  components.map! { |c| c['long_name'] }
  components.reverse.each do |c|
    address << '-' if previous&.match(pattern) && c.match(pattern)
    address << c
    previous = c
  end
  station['address'] = address
  station
end

File: src/ts/geocoding.ts
import axios from 'axios'
import dotenv from 'dotenv'
import { CSVStation } from './model/station'

interface AddressComponent {
  long_name: string,
  short_name: string,
  types: string[],
}

interface GeocodingResult {
  address_components: AddressComponent[],
  formatted_address: string,
}

interface GeocodingResponse {
  status: string,
  results: GeocodingResult[],
}

// 環境変数の読み込み
dotenv.config({ path: 'src/.env.local' })
const API_KEY = process.env.GOOGLE_GEOCODING_API_KEY

if (!API_KEY) {
  throw new Error('GOOGLE_GEOCODING_API_KEY is not set in environment variables')
}

export async function fetchAddress(station: Readonly<CSVStation>): Promise<Pick<CSVStation, 'postal_code' | 'address'>> {
  console.log(`駅の住所を取得中: ${station.name}`)

  try {
    const response = await axios.get<GeocodingResponse>('https://maps.googleapis.com/maps/api/geocode/json', {
      params: {
        latlng: `${station.lat},${station.lng}`,
        key: API_KEY,
        language: 'ja',
      },
    })

    if (response.data.status !== 'OK') {
      throw new Error(`Geocoding API error: ${JSON.stringify(response.data, null, 2)}`)
    }

    const result = response.data.results[0]
    console.log('response:', result)

    // 郵便番号の抽出
    const postalComponents = result.address_components.filter(
      c => c.types.includes('postal_code'),
    )

    if (postalComponents.length !== 1) {
      throw new Error('Failed to extract postal code')
    }

    const dst = {
      postal_code: postalComponents[0].long_name,
      address: '',
    }

    // 住所の構築
    const exception = ['postal_code', 'country', 'bus_station', 'train_station', 'transit_station']
    const addressComponents = result.address_components
      .filter(c => !c.types.some(t => exception.includes(t)))
      .map(c => c.long_name)

    let address = ''
    const numberPattern = /^[0-9０-９]+$/
    let previous: string | null = null

    addressComponents.reverse().forEach(component => {
      if (previous && numberPattern.test(previous) && numberPattern.test(component)) {
        address += '-'
      }
      address += component
      previous = component
    })

    dst.address = address
    console.log('result:', dst)
    return dst
  } catch (error) {
    if (axios.isAxiosError(error)) {
      throw new Error(`Failed to fetch geocoding data: ${error.message}`)
    }
    throw error
  }
} 